### PR TITLE
Basic support for reading data from nspawn containers

### PR DESCRIPTION
**Pull request by lsiudut**
The PR introduces support for connecting to the underlying container using disk socket, relying on /proc/<PID>/root symlink and default socket path. The reason for this approach is that `setns` syscall will not work in multi-threaded apps - using approach when we enter the namespace and run from there would require forking, shutting all threads down and introducing multi-process communication.

The change introduces the following functionality:
- discovery of running machines using systemd API
- global configs are used for now
- auto discovery is going to make the monitord monitor ALL of the underlying machines
- instead of auto discovery `machine_list` may be used with explicit list of machine names

To re-use existing code a new structure, MachineStats, got introduced. It holds metrics only. MonitordStats remains unchanged, as its format determines the output (via Serde).

There's a field for improvement around reading the machine-specific metrics, eg. main OS could be treated as container and handled separately only after full metrics readout. For now, given that already plenty of changes got introduced, I decided to keep the main OS and containers logic separatedd.

Also, future changes may involve per-container configuration.

### CODE REVIEW

File: src/pid1.rs [hunk 1]
```diff
@@ -21,12 +21,12 @@ pub struct Pid1Stats {
     pub tasks: u64,
 }
 
-/// Get procfs info on pid 1 - <https://manpages.debian.org/buster/manpages/procfs.5.en.html>
-pub fn get_pid1_stats() -> anyhow::Result<Pid1Stats> {
+/// Get procfs info on pid 1 - https://manpages.debian.org/buster/manpages/procfs.5.en.html
+pub fn get_pid_stats(pid: i32) -> anyhow::Result<Pid1Stats> {
```
Comment (cooperlees) [inline src/pid1.rs:25]: TIL linux PID table is still i32 limited ...

File: src/machines.rs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+}
+pub async fn update_machines_stats(
+    config: crate::config::Config,
+    connection: zbus::Connection,
+    locked_monitord_stats: Arc<RwLock<MonitordStats>>,
+) -> anyhow::Result<()> {
+    let locked_machine_stats: Arc<RwLock<MachineStats>> =
+        Arc::new(RwLock::new(MachineStats::default()));
+
+    for (machine, leader) in get_machines(&connection, &config).await?.into_iter() {
```
Comment (cooperlees) [inline src/machines.rs:38]: Took me a bit to understand this / what it is - Does this make it clearer?

Ignore me here if `'leader'` is accepted terminology in systemd machine terms and documentation. Would be my lack of understanding if so.
```suggestion
    for (machine, leader_pid) in get_machines(&connection, &config).await?.into_iter() {
```

**Review by cooperlees — APPROVED** [review]
LGTM and thanks. I actually like just collecting a machine stats for main host and inserting into the main struct like we do. Let's leave it like that ... Mainly cause people might already be expecting that etc. and I don't see a huge benefit to combine.
- Also clearer which is the bare metal main OS stats IMO ...

Just have some questions / nits / suggestions. Nothing major. Happy to merge as is, but just want to get your thoughts on a few things.

File: src/machines.rs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+pub async fn get_machines(
+    connection: &zbus::Connection,
+    config: &crate::config::Config,
+) -> Result<HashMap<String, u32>, zbus::Error> {
+    let c = crate::dbus::zbus_machines::ManagerProxy::new(connection).await?;
+    let mut results = HashMap::<String, u32>::new();
+
+    let machines = c.list_machines().await?;
+
+    for machine in machines.iter().filter(|c| &c.1 == "container") {
```
[the remark this answers is not in the document]
Comment (lsiudut) [reply src/machines.rs:19]: Actually it was mu thought. Apparently we could just create a Serde object for it and let it serialize into it. However it seems that there's some code involved to make it strict enough for Rust.

File: src/machines.rs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    let locked_machine_stats: Arc<RwLock<MachineStats>> =
+        Arc::new(RwLock::new(MachineStats::default()));
+
+    for (machine, leader) in get_machines(&connection, &config).await?.into_iter() {
+        debug!(
+            "Collecting container: machine: {} leader: {}",
+            machine, leader
+        );
+        let container_address =
+            format!("unix:path=/proc/{}/root/run/dbus/system_bus_socket", leader);
```
[the remark this answers is not in the document]
Comment (lsiudut) [reply src/machines.rs:44]: Definitely. I wanted to use one from the global config, but that would require parsing the socket string and amending it, I decided not to squeeze it into this batch of already significant changes.

File: src/pid1.rs (hunk 1 above)
Comment (lsiudut) [reply src/pid1.rs:25]: The last time I checked PID_MAX_LIMIT was 2^22, so well within i32. I wonder if there's a use case for negatives here though.

File: src/machines.rs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+        if config.machines.auto_discover || config.machines.machines_list.contains(&machine.0) {
+            results.insert(machine.0.to_string(), leader);
+        }
+    }
+
+    return Ok(results);
+}
+pub async fn update_machines_stats(
+    config: crate::config::Config,
```
[the remark this answers is not in the document]
Comment (lsiudut) [reply src/machines.rs:31]: Yeah. Changing that causes issues with lifetime of the config object from `lib`. Simply following compiler suggestions makes it super weird, especially that other collector methods are also accepting cloned object.

File: src/lib.rs [hunk 6]
```diff
@@ -140,7 +166,19 @@ pub async fn stat_collector(
             }
         }
 
+        {
+            // Update monitord stats with machine stats
+            let mut monitord_stats = locked_monitord_stats.write().await;
+            let machine_stats = locked_machine_stats.read().await;
+            monitord_stats.pid1 = machine_stats.pid1.clone();
+            monitord_stats.networkd = machine_stats.networkd.clone();
+            monitord_stats.system_state = machine_stats.system_state.clone();
```
[the remark this answers is not in the document]
Comment (lsiudut) [reply src/lib.rs:175]: To be perfectly honest I'm not sure what to do with this one... :')

[the remark this answers is not in the document]
Comment (cooperlees) [reply src/lib.rs:175]: Ignore if clippy does not care I say for now

**Pull request merged**